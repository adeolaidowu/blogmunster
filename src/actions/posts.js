import database from "../firebase/firebase";

//ADD_POST
export const addPost = (post) => ({
  type: "ADD_POST",
  post,
});

export const startAddPost = (postData = {}) => {
  return (dispatch) => {
    const {
      title = "",
      content = "",
      imageLink = "",
      createdAt = 0,
    } = postData;
    const post = { title, content, imageLink, createdAt };
    return database
      .ref("posts")
      .push(post)
      .then((ref) => {
        dispatch(
          addPost({
            id: ref.key,
            ...post,
          })
        );
      });
  };
};

//EDIT_POST
export const editPost = (id, updates) => ({
  type: "EDIT_POST",
  id,
  updates,
});

export const startEditPost = (id, updates) => {
  return (dispatch) => {
    dispatch(editPost(id, updates));
    database.ref(`posts/${id}`).update(updates);
  };
};

//DELETE_POST
export const deletePost = ({ id } = {}) => ({
  type: "DELETE_POST",
  id,
});

export const startDeletePost = ({ id } = {}) => {
  return (dispatch) => {
    database
      .ref(`posts/${id}`)
      .remove()
      .then(() => {
        dispatch(deletePost({ id }));
      });
  };
};

//SET_POSTS -- This action fetches data from db to display when page is visited
export const setPosts = (posts) => ({
  type: "SET_POSTS",
  posts,
});

export const startSetPosts = () => {
  return (dispatch) => {
    return database
      .ref("posts")
      .once("value")
      .then((snapshot) => {
        const posts = [];
        snapshot.forEach((childSnapshot) => {
          posts.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setPosts(posts));
      });
  };
};
