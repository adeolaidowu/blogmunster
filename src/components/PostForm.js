import React, { Component } from "react";
import moment from "moment";
import { storage } from "../firebase/firebase";

class PostForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.post ? props.post.title : "",
      content: props.post ? props.post.content : "",
      createdAt: props.post ? moment(props.post.createdAt) : moment(),
      image: null,
      imageLink: props.post ? props.post.imageLink : "",
      error: "",
      fileInputChanged: false,
    };
  }

  // handleChange = (e) => {
  //   this.setState(() => {
  //     const id = e.target.id;
  //     const value = e.target.value;
  //     return { id: value };
  //   });
  // };..

  handleTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };

  handleContentChange = (e) => {
    const content = e.target.value;
    this.setState(() => ({ content }));
  };

  handleSubmit = (e) => {
    e.preventDefault();
    if (!this.state.title || !this.state.content) {
      this.setState(() => ({
        error: "Please provide title and content",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        title: this.state.title,
        content: this.state.content,
        createdAt: this.state.createdAt.valueOf(),
        imageLink: this.state.imageLink,
      });
    }
  };

  handleFileInput = () => {
    const fileInput = document.getElementById("imageInput");
    fileInput.click();
  };

  // handleImageChange = (e) => {
  //   if (e.target.files[0]) {
  //     const image = e.target.files[0];
  //     this.setState(() => ({
  //       image,
  //     }));
  //     const uploadTask = storage.ref(`images/${image.name}`).put(image);
  //     uploadTask.on(
  //       "state_changed",
  //       (snapshot) => {
  //         const progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
  // console.log('Upload is ' + progress + '% done');;
  //       },
  //       (error) => {
  //         console.log("error", error);
  //       },
  //       () => {
  //         storage
  //           .ref("images")
  //           .child(image.name)
  //           .getDownloadURL()
  //           .then((url) => {
  //             this.setState(() => ({
  //               imageLink: url,
  //             }));
  //             console.log("firebase url is", url);
  //           });
  //       }
  //     );
  //   }
  // };

  handleImageChange = (e) => {
    const image = e.target.files[0];
    this.setState(() => ({
      image,
    }));
    console.log(e.target.value);
  };

  handleImageUpload = () => {
    const uploadTask = storage
      .ref()
      .child(`images/${this.state.image.name}`)
      .put(this.state.image);
    uploadTask.on(
      "state_changed",
      (snapshot) => {
        const progress =
          (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
        console.log("Upload is " + progress + "% done");
      },
      (error) => {
        console.log("error", error);
      },
      () => {
        uploadTask.snapshot.ref.getDownloadURL().then((downloadURL) => {
          this.setState(() => ({
            imageLink: downloadURL,
          }));
          console.log("File available at", downloadURL);
        });
      }
    );
  };

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        {this.state.error && <p>{this.state.error}</p>}
        <input
          type="text"
          placeholder="Post title"
          id="title"
          value={this.state.title}
          onChange={this.handleTitleChange}
        />
        <textarea
          placeholder="Post content"
          id="content"
          value={this.state.content}
          onChange={this.handleContentChange}
        ></textarea>
        <input
          type="file"
          id="imageInput"
          //hidden="hidden"
          onChange={this.handleImageChange}
        />
        <button type="button" onClick={this.handleFileInput}>
          Add Image
        </button>
        <button type="button" onClick={this.handleImageUpload}>
          Upload Image
        </button>
        <button>Post</button>
      </form>
    );
  }
}

export default PostForm;
