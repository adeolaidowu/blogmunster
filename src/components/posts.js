import { v4 as uuidv4 } from "uuid";

//ADD_POST
export const addPost = ({
  title = "",
  content = "",
  imageLink = "",
  createdAt = 0,
} = {}) => ({
  type: "ADD_POST",
  post: {
    id: uuidv4(),
    title,
    content,
    imageLink,
    createdAt,
  },
});

//EDIT_POST
export const editPost = (id, updates) => ({
  type: "EDIT_POST",
  id,
  updates,
});

//DELETE_POST
export const deletePost = ({ id } = {}) => ({
  type: "DELETE_POST",
  id,
});
