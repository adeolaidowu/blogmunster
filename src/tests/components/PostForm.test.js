import React from "react";
import PostForm from "../../components/PostForm";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import posts from "../fixtures/posts";

test("should render correctly with no post prop", () => {
  const wrapper = shallow(<PostForm />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should render component with post prop", () => {
  const wrapper = shallow(<PostForm post={posts[1]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should update state on title change", () => {
  const title = "express";
  const wrapper = shallow(<PostForm />);
  wrapper.find("input").simulate("change", {
    target: { value: title },
  });
  expect(wrapper.state("title")).toBe(title);
});

test("should render error on submit of empty form", () => {
  const wrapper = shallow(<PostForm />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {},
  });
  expect(wrapper.state("error")).toBeTruthy();
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should update state on content change", () => {
  const content = "about programming";
  const wrapper = shallow(<PostForm />);
  wrapper.find("textarea").simulate("change", {
    target: { value: content },
  });
  expect(wrapper.state("content")).toBe(content);
});

test("should call onsubmit prop with valid form data", () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<PostForm post={posts[2]} onSubmit={onSubmitSpy} />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {},
  });
  expect(wrapper.state("error")).toBe("");
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    title: posts[2].title,
    content: posts[2].content,
    createdAt: posts[2].createdAt,
  });
});
