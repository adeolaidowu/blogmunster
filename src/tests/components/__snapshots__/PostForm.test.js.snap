// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component with post prop 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    id="title"
    onChange={[Function]}
    placeholder="Post title"
    type="text"
    value="anime"
  />
  <textarea
    id="content"
    onChange={[Function]}
    placeholder="Post content"
    value="about anime"
  />
  <button>
    Post
  </button>
</form>
`;

exports[`should render correctly with no post prop 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    id="title"
    onChange={[Function]}
    placeholder="Post title"
    type="text"
    value=""
  />
  <textarea
    id="content"
    onChange={[Function]}
    placeholder="Post content"
    value=""
  />
  <button>
    Post
  </button>
</form>
`;

exports[`should render error on submit of empty form 1`] = `
<form
  onSubmit={[Function]}
>
  <p>
    Please provide title and content
  </p>
  <input
    id="title"
    onChange={[Function]}
    placeholder="Post title"
    type="text"
    value=""
  />
  <textarea
    id="content"
    onChange={[Function]}
    placeholder="Post content"
    value=""
  />
  <button>
    Post
  </button>
</form>
`;
